Hw03 test framework

forprogr opened the pull request.

forprogr committed:
- настройка maven для нового модуля

forprogr committed:
Otus. Домашнее задание #3
Реализация задачи:
1) создать три аннотации - @Test, @Before, @After.
2) Создать класс-тест, в котором будут методы, отмеченные аннотациями.
3) Создать "запускалку теста". На вход она должна получать имя класса с тестами.
4) "Запускалка" должна в классе-тесте найти и запустить методы, отмеченные аннотациями.
5) Алгоримт запуска должен быть такой:
метод Before
методы Test
метод After
для каждой такой "тройки" надо создать СВОЙ объект класса-теста.
6) Исключение в одном тесте не должно прерывать весь процесс тестирования.

+ @Before переделал @BeforeTest, @After переделал в @AfterTest
+ Добавил аннотации @BeforeAll, @AfterAll

stvort commented on hw03-testing-framework/src/main/java/ru/forprogr/hw/hw03testingframework/TestsRunner.java:
```diff
@@ -0,0 +1,120 @@
+	private Class<?> testedClass;
+	private String testedClassName;
+
+	private ArrayList<Method> methodsBeforeAll;
```
Лучше по интерфейсу ```List<Method>```

stvort commented on hw03-testing-framework/src/main/java/ru/forprogr/hw/hw03testingframework/TestsRunner.java:
```diff
@@ -0,0 +1,120 @@
+	}
+
+	private void runBeforeAfterAllMethods(ArrayList<Method> p_methodsAll){
+		MethodsExecutor executor  = new MethodsExecutor(testedClass);
```
Методы BeforeAll и AfterAll статические. Им не обязательно нужен экземпляр для запуска.

stvort commented on hw03-testing-framework/src/main/java/ru/forprogr/hw/hw03testingframework/TestsRunner.java:
```diff
@@ -0,0 +1,120 @@
+	}
+
+	private void runTestMethods(){
+		runBeforeAfterAllMethods(methodsBeforeAll);
```
Лучше чтобы исключения обрабатывались таким образом, что если AeforeAll вылетел то выполнился только AfterAll, а остальное нет. То же с BeforeTest. Если он вылетел то не выполняем тесты, а AfterTest и AfterTest выполняем

stvort commented on hw03-testing-framework/src/main/java/ru/forprogr/hw/hw03testingframework/AfterAll.java:
```diff
@@ -0,0 +1,16 @@
+
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.METHOD)
+public @interface AfterAll {
```
Классы лучше сгруппировать по пакетам. Main в корневом, остальное в подпакетах

stvort commented on hw03-testing-framework/src/main/java/ru/forprogr/hw/hw03testingframework/MethodsExecutor.java:
```diff
@@ -0,0 +1,77 @@
+			p_method.invoke(testedObject);
+		} catch (IllegalAccessException | InvocationTargetException e) {
+			addError("Error when execute method: "+p_method.getName());
+			addError(e.toString());
```
e.getCause().toString() будет поинтереснее. По крайней мере для InvocationTargetException

forprogr committed:
- классы перенесены в субпакеты
- методы @BeforeAll, @AfterAll - должны быть статичными, вызов осуществляется как статичных, либо метод не будет запущен
- если BeforeAll вылетел то выполнился только AfterAll, а остальное нет. То же с BeforeTest. Если он вылетел то не выполняем тесты, а AfterTest и AfterTest выполняем.
- везде переделал вызов на List<Method>
- для InvocationTargetException сделан вызов e.getCause().toString() вместо e.toString()
